test/doc/developer/reverse.cpp

```diff
@@ -0,0 +1,194 @@
+
+#include <eve/wide.hpp>
+#include <eve/function/bit_cast.hpp>
+#include <eve/function/store.hpp>
+
+#include <algorithm>
+#include <bit>
+#include <concepts>
+
+#include <type_traits>
+
+#if defined(EVE_HW_X86)
+
+namespace eve_extra {
+namespace _reverse {
+
+template <typename T, typename N, typename ABI>
+eve::wide<T, N, ABI> actual_reverse(eve::wide<T, N, ABI> x) requires(sizeof(T) == 8)
+{
+       if constexpr (std::same_as<ABI, eve::x86_128_>) return {_mm_shuffle_epi32(x.storage(), _MM_SHUFFLE(1, 0, 3, 2))};
+  else if constexpr (std::same_as<ABI, eve::x86_256_>) return {_mm256_permutevar8x32_epi32(x.storage(), _MM_SHUFFLE(3, 2, 1, 0))};
+}
+
+template <typename N>
+constexpr std::uint32_t mm_shuffle() {
+  if constexpr (N() == 4) return _MM_SHUFFLE(0, 1, 2, 3);
+  if constexpr (N() == 2) return _MM_SHUFFLE(3, 2, 0, 1);
+}
+
+template <typename T, typename N, typename ABI>
+eve::wide<T, N, ABI> actual_reverse(eve::wide<T, N, ABI> x) requires(sizeof(T) == 4)
+{
+  auto raw = x.storage();
+
+       if constexpr (std::same_as<ABI, eve::x86_128_>) return {_mm_shuffle_epi32(raw, mm_shuffle<N>())};
+  else if constexpr (std::same_as<ABI, eve::x86_256_>)
+  {
+    using mask_type = eve::wide<int, eve::fixed<8>>;
+    auto mask = mask_type([](int i, int) { return 7 - i; }).storage();
+
+    return {_mm256_permutevar8x32_epi32(raw, mask)};
+  }
+}
+
+inline __m256i reverse_256_using_128_mask(__m256i x, __m128i mask_128)
+{
+  __m256i mask_256 = _mm256_set_m128i(mask_128, mask_128);
+  __m256i each_half = _mm256_shuffle_epi8(x, mask_256);
+
+  return {_mm256_permute4x64_epi64(each_half, _MM_SHUFFLE(1, 0, 3, 2))};
+}
+
+template <typename N>
+__m128i reverse_16_chars_mask()
+{
+  return eve::wide<char, eve::fixed<16>>(
+    [](int i, int) { return i >= N() ? i : N() - 1 - i; }
+  ).storage();
+}
+
+template <typename T, typename N, typename ABI>
+eve::wide<T, N, ABI> actual_reverse(eve::wide<T, N, ABI> x) requires(sizeof(T) == 1)
+{
+  auto raw = x.storage();
+
+       if constexpr (eve::current_api <= eve::sse3)
+  {
+    static constexpr std::size_t size = static_cast<std::size_t>(N());
+
+    alignas(sizeof(T) * size) std::array<T, size> buffer;
```

Comment:
I think it'll be better to just use the lambda constructor to reconstruct a wide in reverse instead of going through memory.

`wide<T> result( [&x](int i, int c) { return x[c-1-i]; }`

?

Reply:
Ok - can do. Will check the codegen though - I suspect we are going to get much bigger binary code. Will have to see.

Reply:
Normally no has it usually use the mmset like intrinsic.